### PR TITLE
VB-2729 Establishment specific contact information, JSON request and responses

### PR DESCRIPTION
**Pull request by AledWEvans**
VB-2729 Establishment specific contact information, Changed new endpoints to accept JSON only for requests and responses, new endpoint have change. inline the https standards and REST

### CODE REVIEW

**Review by danbenton-mojdt — APPROVED** [review]

**Review by tpmcgowan — COMMENTED** [review]
Tested locally with Postman and works fine. Will leave approving the actual Kotlin to someone who understands it (@danbenton-mojdt  / @dhirajshettymoj)!

**Review by danbenton-mojdt — DISMISSED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt [hunk 1]
```diff
@@ -255,4 +222,126 @@ class PrisonService(
...
+
+    contactDetails.webAddress?.let {
+      if (contactDetailsRepository.isWebAddressOrphaned(it)) {
+        webAddressRepository.delete(it)
+      }
+    }
+  }
+
+  private fun haveContactDetailsChanged(updateContactDetailsDto: ContactDetailsDto, contactDetails: ContactDetails): Boolean {
+    return updateContactDetailsDto.type != contactDetails.type ||
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt:342]: .equals after converting to ContactDto?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/PrisonContactDetailsResource.kt [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+            schema = Schema(implementation = ErrorResponse::class),
+          ),
+        ],
+      ),
+    ],
+  )
+  fun getContactDetails(
+    @Schema(description = "Prison ID", example = "MDI", required = true)
+    @PathVariable
+    @Size(max = 12, min = 2)
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/PrisonContactDetailsResource.kt:80]: this has been taken from there code.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt (hunk 1 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt:342]: DTO should not be used to compare two objects that are not necessarily the same.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/dto/ContactDetailsDto.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+@Schema(description = "Contact information for a prison department")
+class ContactDetailsDto(
+  @Schema(description = "Department Type", example = "SOCIAL_VISIT or PRISON", required = true)
+  @field:NotNull
+  val type: DepartmentType,
+  @Schema(description = "email address", example = "aled.evans@moj.gov.uk", required = false)
+  @field:Email
+  val emailAddress: String? = null,
+  @Schema(description = "Phone Number", example = "01348811539", required = false)
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/dto/ContactDetailsDto.kt:19]: no worries this is my parents tel number,... the phone number validator will not allow dummy values, so would rather not give them a value that wont work.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt [hunk 4]
```diff
@@ -159,82 +175,33 @@ class PrisonService(
...
   @PreAuthorize(CLIENT_CAN_MAINTAIN_ADDRESSES)
   fun deleteEmailAddress(prisonId: String, departmentType: DepartmentType, throwNotFound: Boolean = false) {
-    val contactDetails = contactDetailsRepository.getByPrisonIdAndType(prisonId, departmentType)
-    contactDetails?.let {
-      it.emailAddress?.let {
-        val emailAddressToBeDeleted = it.value
-        if (contactDetails.phoneNumber == null) {
-          contactDetailsRepository.delete(contactDetails)
-        }
+    contactDetailsRepository.getByPrisonIdAndType(prisonId, departmentType)?.let { contactDetails ->
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt:193]: have added extra logs

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/dto/ContactDetailsDto.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/dto/ContactDetailsDto.kt:19]: ok. only worry is this code is in the public domain on github

[the remark this answers is not in the document]
Comment (tpmcgowan) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/dto/ContactDetailsDto.kt:19]: Could just put `01234567890` 🤷‍♂️

**Review by dhirajshettymoj — APPROVED** [review]
LGTM - just added a comment against the phone number.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/dto/ContactDetailsDto.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/dto/ContactDetailsDto.kt]: Yeha I dont want my M and D getting spammed,... I have used 2 dummy numbers 01234567890 and 01234567880

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

**Review by danbenton-mojdt — APPROVED** [review]

**Pull request merged**